### PR TITLE
Call gDSFP with the right state in react-test-render

### LINKED ISSUE #13027: Shallow renderer is not handling getDerivedStateFromProps properly
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

After using enzyme's `setState`, `getDerivedStateFromProps` is called with an older version of state.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

```jsx
import { shallow } from 'enzyme';

class Demo extends Component {
  static getDerivedStateFromProps(props, state) {
    return { value: state.value };
  };

  state = { value: 'old' };

  render() {
    return <div />;
  }
}

const wrapper = shallow(<Demo />);
wrapper.setState({ value: 'new' });
assert(wrapper.state().value === 'new'); // this throws
```

While I don't have a standalone demo, I tracked the problem to the `_updateStateFromStaticLifecycle` method of `ReactShallowRenderer`. Inside of it `getDerivedStateFromProps` is called with `this._instance.state`, but it should be called with `this._newState`. This might be related to the change in when `getDerivedStateFromProps` is being called.

**What is the expected behavior?**

```jsx
assert(wrapper.state().value === 'new'); // this doesn't throw
```

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

react-test-renderer@16.4.0
enzyme-adapter-react-16@1.1.1
enzyme@3.3.0

### PR DESCRIPTION
**Pull request by fatfisz**
Fixes #13027.

### CODE REVIEW

File: packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js [hunk 1]
```diff
@@ -915,6 +915,39 @@ describe('ReactShallowRenderer', () => {
     expect(result.props.children).toEqual(3);
   });
 
+  it('should call getDerivedStateFromProps with the state set by setState', () => {
+    class SimpleComponent extends React.Component {
```
Comment (gaearon) [inline packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js]: Is this based on any existing test in other renderers? If not, can you please find an appropriate existing test and then use it as a base?

Comment (fatfisz) [reply packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js]: I'll look for that!

Comment (fatfisz) [reply packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js]: The only test I could find that was related to this was: "should not override state with stale values if prevState is spread within getDerivedStateFromProps" in `react/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js`. I modified it a bit (it involved a parent and a child component, but in the case of shallow rendering this seemed to be a bit too complicated).

**Comment by fatfisz** [comment]
Just wondering, when will you be cutting a new release?

**Comment by gaearon** [comment]
Maybe later this week

**Comment by kamronbatman** [comment]
@gaearon, please cut a new version asap, <3

**Comment by gaearon** [comment]
It was included in 16.4.1.

**Comment by kamronbatman** [comment]
@gaearon, I am still having a similar issue. It might be due to a different reason then (I am not sure at this time).
Thank you!

**Comment by gaearon** [comment]
Can you file a bug with reproducing case please?

**Comment by gaearon** [comment]
Or maybe you're getting confused about which behavior is intended?
There was a bug in 16.3 which you might accidentally have relied on.

See:

https://reactjs.org/blog/2018/05/23/react-v-16-4.html#bugfix-for-getderivedstatefromprops

and

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html

**Pull request merged**